In [14]:
# Only for collab
!pip install langdetect
from google.colab import drive
import pandas as pd
drive.mount("/content/drive")
df = pd.read_csv('/content/drive/MyDrive/full_merged_dataset15.csv', encoding='utf-8')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import pandas as pd
import os
import numpy as np
from termcolor import colored
import seaborn as sns
import matplotlib.pyplot as plt
import html
from langdetect import detect
from nltk.stem import SnowballStemmer
from nltk.stem import WordNetLemmatizer

In [3]:
#df = pd.read_csv('full_merged_dataset15.csv', encoding='utf-8')

In [15]:
df.head()

id                         title                   author  \
0  162275   The 80&rsquo;s Miracle Diet          By Melvin Dixon   
1  162250             All Saints&rsquo;  By Corey Van Landingham   
2  162279  And These Are Just a Few ...          By Melvin Dixon   
3  162248                          ASMR  By Corey Van Landingham   
4  162249                          ASMR  By Corey Van Landingham   

                                             snippet  \
0                      Yours free without the asking   
1    Caravaggio’s face in the sunken pumpkin.   
2  This poem is for the epidemic dead and the liv...   
3                      Why not climb up the mountain   
4                               Hello—Tonight   

                                                link  \
0  https://www.poetryfoundation.org/poetrymagazin...   
1  https://www.poetryfoundation.org/poetrymagazin...   
2  https://www.poetryfoundation.org/poetrymagazin...   
3  https://www.poetryfoundation.org/poetrymagazin...   
4  https://www.poetryfoundation.org/poetrymagazin...   

                                          categories  \
0                     ['Living', 'Health & Illness']   
1                                                 []   
2  ['Living', 'Health & Illness', 'Social Comment...   
3                                                 []   
4                                                 []   

                                                poem  
0  Yours free without the asking Quick delivery v...  
1  Caravaggio’s face in the sunken pumpkin...  
2  This poem is for the epidemic dead and the liv...  
3  Why not climb up the mountain of delight? To t...  
4  Hello—Tonight we’ll trace the st...

We can remove the id, snippet and link as they are useless for the tasks we have to do.

In [16]:
columns_to_drop = ["id", "snippet", "link"]
df.drop(columns=columns_to_drop, inplace=True)

Codificar bien el texto

In [25]:
df.head()

title                   author  \
0   The 80&rsquo;s Miracle Diet          By Melvin Dixon   
1             All Saints&rsquo;  By Corey Van Landingham   
2  And These Are Just a Few ...          By Melvin Dixon   
3                          ASMR  By Corey Van Landingham   
4                          ASMR  By Corey Van Landingham   

                                          categories  \
0                     ['Living', 'Health & Illness']   
1                                                 []   
2  ['Living', 'Health & Illness', 'Social Comment...   
3                                                 []   
4                                                 []   

                                                poem  
0  Yours free without the asking Quick delivery v...  
1  Caravaggio’s face in the sunken pumpkin...  
2  This poem is for the epidemic dead and the liv...  
3  Why not climb up the mountain of delight? To t...  
4  Hello—Tonight we’ll trace the st...

In [26]:
import re
import nltk

def check_nltk_packages():
  packages = ['punkt','stopwords','omw-1.4','wordnet']

  for package in packages:
    try:
      nltk.data.find('tokenizers/' + package)
    except LookupError:
      nltk.download(package)
check_nltk_packages()

try:
  import lxml
except ModuleNotFoundError:
  %pip install lxml

try:
  import contractions
except ModuleNotFoundError:
  %pip install contractions
  import contractions

from bs4 import BeautifulSoup
import re

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [27]:

def separate_language(df, text_column):
    """
    Separate observations in a DataFrame based on the language detected in a text column.

    Args:
        df (pandas.DataFrame): Input DataFrame.
        text_column (str): Name of the column containing text data.

    Returns:
        tuple: A tuple containing two DataFrames. The first DataFrame contains observations in English,
               and the second DataFrame contains observations in other languages.
    """
    # Initialize lists to store data
    english_data = []
    other_language_data = []

    # Loop through each row in the dataset
    for index, row in df.iterrows():
        text = row[text_column]
        try:
            # Detect language of the text
            lang = detect(text)

            # Append entire row data to respective lists
            if lang == 'en':
                english_data.append(row)
            else:
                other_language_data.append(row)
        except:
            # Handle exceptions in language detection (e.g., short texts)
            pass

    # Create dataframes from lists
    english_df = pd.DataFrame(english_data)
    other_language_df = pd.DataFrame(other_language_data)

    return english_df, other_language_df

In [28]:
# Call the function to separate language
english_df, other_language_df = separate_language(df, 'poem')

In [29]:
# i want to check the size of both dfs
print(english_df.shape)
print(other_language_df.shape)

(44862, 4)
(426, 4)


Only 426 poems are not in english and we know that most of them have been translated to english. Note that this value changes as expected

In [30]:
english_df.head()


title                   author  \
0   The 80&rsquo;s Miracle Diet          By Melvin Dixon   
1             All Saints&rsquo;  By Corey Van Landingham   
2  And These Are Just a Few ...          By Melvin Dixon   
3                          ASMR  By Corey Van Landingham   
4                          ASMR  By Corey Van Landingham   

                                          categories  \
0                     ['Living', 'Health & Illness']   
1                                                 []   
2  ['Living', 'Health & Illness', 'Social Comment...   
3                                                 []   
4                                                 []   

                                                poem  
0  Yours free without the asking Quick delivery v...  
1  Caravaggio’s face in the sunken pumpkin...  
2  This poem is for the epidemic dead and the liv...  
3  Why not climb up the mountain of delight? To t...  
4  Hello—Tonight we’ll trace the st...

In [31]:
other_language_df.head()

title              author  \
10              Fosfor  By Kacper Bartczak   
29  Pieśń tego  By Kacper Bartczak   
32       Prawda obrazu  By Kacper Bartczak   
35   regenportr&auml;t       By Jan Wagner   
37              reifen       By Jan Wagner   

                                        categories  \
10                                              []   
29                                              []   
32                                              []   
35                                      ['Nature']   
37  ['Social Commentaries', 'Cities & Urban Life']   

                                                 poem  
10  w terrarium jesteÅ› czarna skrzyneczka rzec...  
29  perło świata wymiocino wieprzu jestem ...  
32  Kobieta naprzeciwko czyta „Największ...  
35  erscheint als handkuß, als sanfter stubser,...  
37  dort, wo die stadt versickert, beim bahndamm, ...

### Save the english DataFrame and the rest of the data

In [43]:
english_df.to_csv('english_df.csv', encoding='utf-8', index=False) # CAREFUL, you can overwrite your previous dataset, UPDATE the name when used
#other languages dataset can not be codified and as we are not going to use it we dont care about it
other_language_df.to_csv('other_language_df.csv',encoding='utf-8', index=False) # CAREFUL, you can overwrite your previous dataset, UPDATE the name when used


# Small Case Text Preprocessing

In [44]:
import numpy as np

df = pd.read_csv('english_df.csv', encoding='utf-8')

# Drop rows with missing values in the 'poem' column
df = df.dropna(subset=['poem'])

In [45]:
df.head(5)

title                   author  \
0   The 80&rsquo;s Miracle Diet          By Melvin Dixon   
1             All Saints&rsquo;  By Corey Van Landingham   
2  And These Are Just a Few ...          By Melvin Dixon   
3                          ASMR  By Corey Van Landingham   
4                          ASMR  By Corey Van Landingham   

                                          categories  \
0                     ['Living', 'Health & Illness']   
1                                                 []   
2  ['Living', 'Health & Illness', 'Social Comment...   
3                                                 []   
4                                                 []   

                                                poem  
0  Yours free without the asking Quick delivery v...  
1  Caravaggio’s face in the sunken pumpkin...  
2  This poem is for the epidemic dead and the liv...  
3  Why not climb up the mountain of delight? To t...  
4  Hello—Tonight we’ll trace the st...

In [46]:
print(df.shape)


(44862, 4)


As the poems were not well-encoded we have to substitude the bad encoding to an apostrophe (also some other characters), and expand the contractions.

In [47]:
# Pre-compile the patterns for efficiency
url_pattern = re.compile(r'https://\S+|www\.\S+')

chars_to_replace_by_apostrophe = re.compile(r'’|’')
chars_to_replace_by_a = re.compile(r'&aacute')

unwanted_chars_pattern = re.compile(r'â€Å|’|—|ۼ|â€Ã|Ã|¬|¢|š|€||âº|â|º|â|Ãâ€|¬ÂºÃƒâ€¦Ã¢â|¬|Å|¾|º|â|º|›Ãâ€')

def wrangle_text(text):

    # Decode HTML entities
    text = html.unescape(text)

    # Remove HTML tags
    text = BeautifulSoup(text, "lxml").text

    # Replace by apostrophe
    text = chars_to_replace_by_apostrophe.sub("'", text)

    # Replace by a
    text = chars_to_replace_by_a.sub('á', text)

    # Fix specific character issues in one go

    text = unwanted_chars_pattern.sub(' ', text)

    # Expand contractions once, after all substitutions
    text = contractions.fix(text)

    return text

In [48]:
poem = df['poem'].iloc[84] # This is the 84th poem
wrangled_poem = wrangle_text(poem)
print(colored('\n============= First poem in corpus =============', 'blue'))
print(poem)
print(colored('\n============= After wrangling result =============', 'green'))
print(wrangled_poem)


============= First poem in corpus =============
I want to believe I’m on my last life. What is nirvana if not a kind of death? In a past life, a stranger asked the Buddha for his children. The Buddha offered him his eyes. Then, pulling the children, screaming, from the shivering rice barrels, he gave them away. Did the children ever forgive him? Did they have to because he was the Buddha? I stay awake, listening to my brother’s breath as he sleeps in my corridor. A 6 am flight, a half-hug in the parking lot. I want one more life. This is the problem with reincarnation: you don’t know if what you’ve lost is lost forever. Unless, I guess, you’re the Buddha. His children escaped, by the way. Hid from their possessor in the cane grass, the silk reeds, and wove their way back home. I think they hugged their mother. I think the Buddha demanded understanding. I had to, he likely said. I wonder who they became in their next life. I wonder if the siblings st

Now we are going to tokenize the poems by words and by sentences

In [49]:
from nltk.tokenize import wordpunct_tokenize
from nltk.tokenize import sent_tokenize


print(colored('\n============= First poem in data =============', 'blue'))
print(wrangled_poem)

# Tokenize the review
poem_tokens = wordpunct_tokenize(wrangled_poem)
# Tokenize the review at the sentence level
poem_tokens_sent = sent_tokenize(wrangled_poem)

print(colored('\n============= First poem (tokens) =============', 'green'))
print(poem_tokens)

print(colored('\n============= First poem (tokens sent level) =============', 'red'))
print(poem_tokens_sent)


============= First poem in data =============
I want to believe I am on my last life. What is nirvana if not a kind of death? In a past life, a stranger asked the Buddha for his children. The Buddha offered him his eyes. Then, pulling the children, screaming, from the shivering rice barrels, he gave them away. Did the children ever forgive him? Did they have to because he was the Buddha? I stay awake, listening to my brother's breath as he sleeps in my corridor. A 6 am flight, a half-hug in the parking lot. I want one more life. This is the problem with reincarnation: you do not know if what you have lost is lost forever. Unless, I guess, you are the Buddha. His children escaped, by the way. Hid from their possessor in the cane grass, the silk reeds, and wove their way back home. I think they hugged their mother. I think the Buddha demanded understanding. I had to, he likely said. I wonder who they became in their next life. I wonder if the siblings stayed together, across this ever-

### Converting into lower case & removing non alphanumeric:

In [50]:
poem_tokens_filtered = [token.lower() for token in poem_tokens if token.isalnum()]

# The isalnum() method returns True if all characters in the string are
# alphanumeric (either alphabets or numbers). If not, it returns False.

print(colored('\n============= First poem (tokens) =============', 'blue'))
print(poem_tokens)

print(colored('\n============= First poem (tokens filtered) =============', 'green'))
print(poem_tokens_filtered)


============= First poem (tokens) =============
['I', 'want', 'to', 'believe', 'I', 'am', 'on', 'my', 'last', 'life', '.', 'What', 'is', 'nirvana', 'if', 'not', 'a', 'kind', 'of', 'death', '?', 'In', 'a', 'past', 'life', ',', 'a', 'stranger', 'asked', 'the', 'Buddha', 'for', 'his', 'children', '.', 'The', 'Buddha', 'offered', 'him', 'his', 'eyes', '.', 'Then', ',', 'pulling', 'the', 'children', ',', 'screaming', ',', 'from', 'the', 'shivering', 'rice', 'barrels', ',', 'he', 'gave', 'them', 'away', '.', 'Did', 'the', 'children', 'ever', 'forgive', 'him', '?', 'Did', 'they', 'have', 'to', 'because', 'he', 'was', 'the', 'Buddha', '?', 'I', 'stay', 'awake', ',', 'listening', 'to', 'my', 'brother', "'", 's', 'breath', 'as', 'he', 'sleeps', 'in', 'my', 'corridor', '.', 'A', '6', 'am', 'flight', ',', 'a', 'half', '-', 'hug', 'in', 'the', 'parking', 'lot', '.', 'I', 'want', 'one', 'more', 'life', '.', 'This', 'is', 'the', 'problem', 'with', 'reincarnation', ':', 'you', 'do', 'not', 'know', 'i

Now we use stemming or lemmatization.

In [51]:
stemmer = SnowballStemmer('english')
wnl = WordNetLemmatizer()

stemmed_poem = [stemmer.stem(el) for el in poem_tokens_filtered]
print(colored('\n============= Stemmed review  =============', 'blue'))
print(stemmed_poem)

lemmatized_poem = [wnl.lemmatize(el) for el in poem_tokens_filtered]
print(colored('\n============= Lemmatized review  =============', 'blue'))
print(lemmatized_poem)

# The difference between both is that the lemmatization process is more complex
# and it involves resolving words to their base or root form.

# Lemmatization: Lemmatization is the process of converting a word to its base form.
# Stemming: Stemming is the process of reducing a word to its root form.

# For example:
# Lemmatization: The word "better" is converted to "good".
# Stemming: The word "better" is converted to "bet".


============= Stemmed review  =============
['i', 'want', 'to', 'believ', 'i', 'am', 'on', 'my', 'last', 'life', 'what', 'is', 'nirvana', 'if', 'not', 'a', 'kind', 'of', 'death', 'in', 'a', 'past', 'life', 'a', 'stranger', 'ask', 'the', 'buddha', 'for', 'his', 'children', 'the', 'buddha', 'offer', 'him', 'his', 'eye', 'then', 'pull', 'the', 'children', 'scream', 'from', 'the', 'shiver', 'rice', 'barrel', 'he', 'gave', 'them', 'away', 'did', 'the', 'children', 'ever', 'forgiv', 'him', 'did', 'they', 'have', 'to', 'becaus', 'he', 'was', 'the', 'buddha', 'i', 'stay', 'awak', 'listen', 'to', 'my', 'brother', 's', 'breath', 'as', 'he', 'sleep', 'in', 'my', 'corridor', 'a', '6', 'am', 'flight', 'a', 'half', 'hug', 'in', 'the', 'park', 'lot', 'i', 'want', 'one', 'more', 'life', 'this', 'is', 'the', 'problem', 'with', 'reincarn', 'you', 'do', 'not', 'know', 'if', 'what', 'you', 'have', 'lost', 'is', 'lost', 'forev', 'unless', 'i', 'guess', 'you', 'are', 'the', 'buddha', 'his', 'children', 'es

# Preprocessing of the whole dataset

Now we have added a stage where we remove stopwords

In [54]:
import numpy as np

df = pd.read_csv('english_df.csv', encoding='utf-8')

# Drop rows with missing values in the 'poem' column
df = df.dropna(subset=['poem'])

In [55]:
print(df.shape)

(44862, 4)


In [56]:
df2 = df.copy()
df2['poem'] = df['poem'].apply(lambda x: wrangle_text(x))

<ipython-input-47-eb8965b45236>:15: MarkupResemblesLocatorWarning: The input looks more like a filename than markup. You may want to open this file and pass the filehandle into Beautiful Soup.
  text = BeautifulSoup(text, "lxml").text


In [61]:
df2.to_csv('data_wrangled.csv',encoding='latin-1', errors='ignore', index=False) # CAREFUL, you can overwrite your previous dataset, UPDATE the name when used

In [64]:
from nltk.corpus import stopwords
# Initialize NLTK resources
stopwords_en = set(stopwords.words('english'))
wnl = WordNetLemmatizer()


def prepare_data(text, tool='lemmatize'):

    # Decode HTML entities
    text = html.unescape(text)

    # Wrangle text
    text = wrangle_text(text)

    # Tokenize text
    tokens = wordpunct_tokenize(text)

    # Filter tokens
    tokens_filtered = [token.lower() for token in tokens if token.isalnum()]

    if tool == 'stem':
        # Stem tokens
        tokens_filtered = [stemmer.stem(el) for el in tokens_filtered]
    elif tool == 'lemmatize':
        # Lemmatize tokens
        tokens_filtered = [wnl.lemmatize(el) for el in tokens_filtered]
    else:
        raise ValueError('Invalid tool. Please select either "stem" or "lemmatize".')

    # Remove stopwords
    clean_tokens = [el for el in tokens_filtered if el not in stopwords_en]

    return clean_tokens

### Processing of the poems column

In [66]:
# Now lets use the original text and check it:
print(colored('\n============= First review in corpus =============', 'blue'))
poem = df2['poem'].iloc[84]
print(poem)

# Prepare the data
clean_tokens = prepare_data(poem, tool = "lemmatize")
print(colored('\n============= Clean tokens =============', 'green'))
print(clean_tokens)



============= First review in corpus =============
I want to believe I am on my last life. What is nirvana if not a kind of death? In a past life, a stranger asked the Buddha for his children. The Buddha offered him his eyes. Then, pulling the children, screaming, from the shivering rice barrels, he gave them away. Did the children ever forgive him? Did they have to because he was the Buddha? I stay awake, listening to my brother's breath as he sleeps in my corridor. A 6 am flight, a half-hug in the parking lot. I want one more life. This is the problem with reincarnation: you do not know if what you have lost is lost forever. Unless, I guess, you are the Buddha. His children escaped, by the way. Hid from their possessor in the cane grass, the silk reeds, and wove their way back home. I think they hugged their mother. I think the Buddha demanded understanding. I had to, he likely said. I wonder who they became in their next life. I wonder if the siblings stayed together, across this e

In [67]:
# Apply the prepare_data function to clean the poems
df2['clean_tokens'] = df2['poem'].apply(lambda x: prepare_data(x, tool='lemmatize'))

<ipython-input-47-eb8965b45236>:15: MarkupResemblesLocatorWarning: The input looks more like a filename than markup. You may want to open this file and pass the filehandle into Beautiful Soup.
  text = BeautifulSoup(text, "lxml").text


In [68]:
df2.head()

title                   author  \
0   The 80&rsquo;s Miracle Diet          By Melvin Dixon   
1             All Saints&rsquo;  By Corey Van Landingham   
2  And These Are Just a Few ...          By Melvin Dixon   
3                          ASMR  By Corey Van Landingham   
4                          ASMR  By Corey Van Landingham   

                                          categories  \
0                     ['Living', 'Health & Illness']   
1                                                 []   
2  ['Living', 'Health & Illness', 'Social Comment...   
3                                                 []   
4                                                 []   

                                                poem  \
0  Yours free without the asking Quick delivery v...   
1  Caravaggio's face in the sunken pumpkin. Bulbs...   
2  This poem is for the epidemic dead and the liv...   
3  Why not climb up the mountain of delight? To t...   
4  Hello Tonight we will trace the static bough, ...   

                                        clean_tokens  
0  [free, without, asking, quick, delivery, via, ...  
1  [caravaggio, face, sunken, pumpkin, bulb, rott...  
2  [poem, epidemic, dead, living, remember, neigh...  
3  [climb, mountain, delight, world, thin, meridi...  
4  [hello, tonight, trace, static, bough, tempora...

In [69]:
# remove the column 'poem'
df2 = df2.drop(columns=['poem'])

In [70]:
print(df.shape)

(44862, 4)


In [71]:
df2.to_csv('data_poems_preprocessed.csv', encoding='latin-1',errors='ignore', index=False) # CAREFUL, you can overwrite your previous dataset, UPDATE the name when used

In [72]:
df2.head(10)

title                   author  \
0                        The 80&rsquo;s Miracle Diet          By Melvin Dixon   
1                                  All Saints&rsquo;  By Corey Van Landingham   
2                       And These Are Just a Few ...          By Melvin Dixon   
3                                               ASMR  By Corey Van Landingham   
4                                               ASMR  By Corey Van Landingham   
5                                     Autumn Leaving          By Melvin Dixon   
6                                             Babble         By Nica Giromini   
7                            Chiaroscuro Spring time        By Cyrus Cassells   
8                                           Currency         By Nica Giromini   
9  Everyday around the world a woman is pulled in...       By Krista Franklin   

                                          categories  \
0                     ['Living', 'Health & Illness']   
1                                                 []   
2  ['Living', 'Health & Illness', 'Social Comment...   
3                                                 []   
4                                                 []   
5                                                 []   
6                                                 []   
7        ['Arts & Sciences', 'Painting & Sculpture']   
8                                                 []   
9                       ['Types/Modes', 'Ekphrasis']   

                                        clean_tokens  
0  [free, without, asking, quick, delivery, via, ...  
1  [caravaggio, face, sunken, pumpkin, bulb, rott...  
2  [poem, epidemic, dead, living, remember, neigh...  
3  [climb, mountain, delight, world, thin, meridi...  
4  [hello, tonight, trace, static, bough, tempora...  
5  [didier, vermont, leaf, glass, hand, working, ...  
6  [say, land, remembered, muddy, rush, slows, ne...  
7  [galleria, nazionale, arte, antica, rome, wa, ...  
8  [net, headfirst, prairie, spring, seeing, noth...  
9  [kara, walker, blue, â, 2020, everyday, brown,...